### PR TITLE
SSLHostConfig implementation for Connector

### PR DESCRIPTION
**Pull request by thanosploumis**
@jstefl Here is my SSLHostConfig inner element implementation for Connector :)

### CODE REVIEW

**Comment by thanosploumis** [comment]
Also wanted to notice that `setSSLHostConfigTrustoreFile and setSSLHostConfigTrustorePath ` are temporary names (there was conflict naming) until we discuss if we will remove old element configurations[1] and replace with new or just add new implementation and leave old as it is (with prettier naming)

[1] https://tomcat.apache.org/tomcat-9.0-doc/config/http.html#SSL_Support_-_Connector_-_NIO_and_NIO2_(deprecated)

File: core/src/main/groovy/noe/tomcat/configure/ConnectorAttributesTransformer.groovy [hunk 1]
```diff
@@ -178,8 +179,19 @@ class ConnectorAttributesTransformer {
         node.appendNode("UpgradeProtocol", ['className': connector.getUpgradeProtocol().getClassName()])
       }
 
+      if (connector.getSSLHostConfig() != null) {
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorAttributesTransformer.groovy:182]: It seems to me, that Groovy has "build-in" good tools for manipulation w/ XMLs. Do we really need a new dependency?
Wouldn't smth similar to ` node.appendNode("SSLHostConfig", connector.getSSLHostConfig() as Map)`  work?

However, my biggest concern would - with this "shortcut" we would lose the very basic validation -isn't it dangerous? Will it not generate unexpected data?
Additionally, it is a bit different concept for 1 element only - consistency is what makes code easy to understand quickly.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy [hunk 2]
```diff
@@ -124,29 +124,33 @@ class ConnectorConfiguratorTomcat {
...
   private void updateExistingConnector(Node connector, Node newConnector) {
-    // update attributes
+    // update connector attributes
     newConnector.attributes().each { attribute ->
       connector.@"${attribute.key}" = attribute.value
     }
 
-    // update nodes
     newConnector.each { Node newSubelement ->
-      String UpgradeProtocol = ConnectorUpgradeProtocolTomcat.ELEMENT_NAME
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:134]: IIUC, we are losing the configuration of UpgradeProtocol, no?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy [hunk 3]
```diff
@@ -124,29 +124,33 @@ class ConnectorConfiguratorTomcat {
...
+  /**
+   *
+   * @param connector
+   * @param newSubElement
+   *
+   * Search for existing inner element to remove and replace with new one
+   * Note: SSLHostConfig inner elements can exist more than one
+   */
+  private void updateInnerElements(Node connector, Node newSubElement) {
+    if (connector.find { it.name() == newSubElement.name() } != null) {
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:146]: What about already existing elements? The methods are called '*update*' ?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy [hunk 4]
```diff
@@ -0,0 +1,110 @@
+package noe.tomcat.configure
+
+/**
+ * Abstraction for SSLHostConfig tag of either Secure or Non-Secure Connector in Tomcat server.xml.
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:4]: The term 'element' is used across the code (instead of tag).

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        return this
+    }
+
+    public ConnectorSSLHostConfigTomcat setTruststoreType(String truststoreType) {
+        this.truststoreType = truststoreType
+        return this
+    }
+
+
+}
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:110]: Please configure your ide to:
 * add new line EOF
 * to have same indentation as rest of the NOE (2 spaces indentation).

File: testsuite/src/test/groovy/noe/tomcat/BindingsTomcatConfiguratorIT.groovy [hunk 6]
```diff
@@ -385,6 +386,43 @@ abstract class BindingsTomcatConfiguratorIT extends TomcatTestAbstract {
...
+    String certPath = "path/to/cert"
+    String tType = "ttype"
+    String tFile = "tfile"
+    String tPass = "tpasswd"
+    String tProvider = "tprovider"
+    String protocols = "protocols"
+    String sslProtocol = "sslprotocol"
+    String ciphers = "several_ciphers"
+
+    Object sslHostConfObj = new ConnectorSSLHostConfigTomcat().setCaCertificateFile(certFile).setCaCertificatePath(certPath)
```
Comment (jstefl) [inline testsuite/src/test/groovy/noe/tomcat/BindingsTomcatConfiguratorIT.groovy:402]: please set each property configuration per line, to be the code more readable

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy [hunk 7]
```diff
@@ -124,29 +124,33 @@ class ConnectorConfiguratorTomcat {
...
+      Node oldElement = (Node) connector.find { it.name() == newSubElement.name()}
+      if (oldElement.name() != "SSLHostConfig") {
+        connector.remove(oldElement)
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:149]: Remove all subelements but `SSLHostConfig` ? So if somebody has some custom configuration (reproducer provided by customer) with upgradeProdocal set already and we want to add some configuration wrt SSLHostConfig, will the upgradeProtocal configuration disappear automagically? Or am I missing something?

**Review by jstefl — COMMENTED** [review]
Thank you for your contribution and time! 
I skimmed the code and left some comments there.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 2 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:134]: No UpgradeProtocol configuration remains the same and working properly

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 3 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:146]: I noticed cannot be done proper update to inner elements of Connector and is not needed. If someone for example wants to update UpgradeProtocol or SSLHostConfig will write the element with the desired attributes. Then the above function will remove the specific inner element and replace it with the new desired.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy (hunk 4 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:4]: Nice suggestion. Missed it :) Will replace it asap

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 7 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:149]: As the new configuration will have SSLHostConfig element, the UpgradeProtocol will stand its ground.The new SSLHostConfig element will be added along with the old. The removing is not related with SSLHostConfig element because someone might want to have multiple SSLHostConfig elements inside a connector.I have seen configurations in which it's used like that. Am i right?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 3 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:146]: Let me explain it

Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:146]: With conf like: 

`new TomcatConfigurator(tomcat)`
  `.httpsConnector(new SecureHttpConnectorTomcat()`
  `.setPort(8443)`
  `.setUpgradeProtocolToHttp2Protocol()`

Now will add/update the connector with SSLHostConfig element leaving UpgradeProtocol as is:
`Object map = new ConnectorSSLHostConfigTomcat().setCaCertificateFile("server.key")`

`new TomcatConfigurator(tomcat)`
  `.httpsConnector(new SecureHttpConnectorTomcat()`
  `.setSSLHostConfig(map))`

the result will be :

`<Connector port="8080" secure="true" SSLEnabled="true" clientAuth="false" sslEnabledProtocols="false">`
`      <UpgradeProtocol className="org.apache.coyote.http2.Http2Protocol"/>`
`      <SSLHostConfig caCertificateFile="server.key"/>`

Keep in mind that updateInnerElements function looking inside newSubElement node which is the new upcoming node.Matches the inner element with the same name.If it's present, removing the old and create a new with the newSubElement attributes.
You were doing the same pattern matching on the old function.But updating attributes independently wasnt working too good.I just made it to remove whole element and replace it with the fresh

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 2 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:134]: This Element name was used in order to identify inner element for update.Now i "scan" new element's name and edit the specific one.Leaving the rest connector unchanged

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 3 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:146]: I see your point. What I had in a mind is a use-case that on input is completely non-default server.xml and I was wondering whether is OK to change the whole element in case that only for instance 1 attribute is configured only. 

Seems to me, that it should be OK since the 'updateExistingConnector' is part of private API and the client see only `TomcatConfigurator#httpConnector` or so.

I would recommend documenting this behavior into public API though.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy [hunk 8]
```diff
@@ -124,29 +124,33 @@ class ConnectorConfiguratorTomcat {
...
+    }
+  }
+
+  /**
+   *
+   * @param connector
+   * @param newSubElement
+   *
+   * Search for existing inner element to remove and replace with new one
+   * Note: SSLHostConfig inner elements can exist more than one
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:143]: What implication of this? What anyone should know?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy [hunk 9]
```diff
@@ -124,29 +124,33 @@ class ConnectorConfiguratorTomcat {
...
+   */
+  private void updateInnerElements(Node connector, Node newSubElement) {
+    if (connector.find { it.name() == newSubElement.name() } != null) {
+      Node oldElement = (Node) connector.find { it.name() == newSubElement.name()}
+      if (oldElement.name() != "SSLHostConfig") {
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:148]: Not sure that I follow why SSLHostConfig has special care?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+ * *
+ * IMPORTANT
+ * <ul>
+ *   <li>Not all SSLHostConfig attributes are supported. Only the most used ones.</li>
+ *   <li>It is user responsibility to set values semantically, no validation is performed.</li>
+ * <ul>
+ *
+ * @link https://tomcat.apache.org/tomcat-9.0-doc/config/http.html#SSL_Support_-_SSLHostConfig
+ */
+class ConnectorSSLHostConfigTomcat {
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:16]: Should be also handled element `Certificate` to make this api change? 
https://tomcat.apache.org/tomcat-9.0-doc/ssl-howto.html

Not sure, just asking.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 2 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:134]: ok, please see my last comment below for details.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorAttributesTransformer.groovy (hunk 1 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorAttributesTransformer.groovy:182]: WDYT?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 9 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:148]: There can be multiple SSLHostConfig elements inside one Connector.So if i would try to add second one it would be deleted and replaced from the new one

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 8 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:143]: What do you mean? I used it as an explanation comment for future development.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy (hunk 10 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:16]: Didnt understand your point

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 9 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:148]: > There can be multiple SSLHostConfig elements inside one Connector.

That is something that tomcat supports? How do we address that? How can one configure multiple SSLHostConfig elements inside one Connector form code?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy (hunk 10 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:16]: There is possible configuraiton like: 
````
<Connector
    port="8443"
    protocol="org.apache.coyote.http11.Http11AprProtocol"
    secure="true"
    scheme="https"
    SSLEnabled="true"
  <SSLHostConfig
      caCertificateFile="/path/to/ca.pem"
      certificateVerification="require"
      certificateVerificationDepth="10" >
    <Certificate
        certificateFile="/path/to/ocsp-cert.crt"
        certificateKeyFile="/path/to/ocsp-cert.key" />
  </SSLHostConfig>
````

I am curious whether is possible to skip the element Certificate for now, and we will have implemented features addressing our needs.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 8 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:143]: Ok, I see.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 3 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:146]: > I would recommend documenting this behavior into public API though.

That the server.xml will contain "exactly" what is set in code, and what are the implication.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 9 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:148]: Isn't it discrepancy with https://github.com/web-servers/noe-core/pull/110/files/f4363f09c53f7ec1b3e01b149c423c44b749f7c5#r539279021?

It seems that something we want to set exactly as it is in code and something else means just an addition. I would consder unifying this approach.

File: core/src/main/groovy/noe/tomcat/configure/ConnectorAttributesTransformer.groovy (hunk 1 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorAttributesTransformer.groovy:182]: Did not forget it :) working on this

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 9 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:148]: Although is possible that multiple SSLHostConfig elements can be present, do you want me to cut it down to only one? 
Tomcat has not document about this but i ve noticed many using it [1].

[1] https://wolfpaulus.com/tomcat-multiple-ssl-certificates/

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy (hunk 10 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:16]: Ah, you mean that as there two elements are related together, it would be better to launch them together to this PR right ?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 3 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:146]: Renamed function to replaceInnerElements.Good point Jan !

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 9 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:148]: I wouldn't cut it down, what I am thinking is that the code shouldn't just add a new element because we don't know whether the client doesn't want to just update existing.

Since the attribute hostname seems as key, I would recommend considering
 * check if the element with key exists, 
     YES -> update it, 
     NO -> add new

WDYT?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy (hunk 10 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:16]: Honestly, I don't know the details. If the use-case we have does not need this element I would consider skipping it ATM,, if it is needed, let's implement it.

Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:16]: I implemented it and is pushed in order to take a look :)

File: core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy (hunk 9 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy:148]: TBH first time I see hostname as attribute.I just shared the link as evidence of multiple SSLHostConfig usability.So I have nothing like key in order to follow your logic.

File: core/src/main/groovy/noe/tomcat/configure/SecureHttpConnectorTomcat.groovy [hunk 11]
```diff
@@ -44,6 +44,7 @@ public class SecureHttpConnectorTomcat extends ConnectorTomcatAbstract<SecureHtt
 
   // Inner elements
   private ConnectorUpgradeProtocolTomcat upgradeProtocol
+  private ConnectorSSLHostConfigTomcat[] SSLHostConfigs
```
Comment (jstefl) [inline core/src/main/groovy/noe/tomcat/configure/SecureHttpConnectorTomcat.groovy:47]: Can we specify it as List<ConnectorSSLHostConfigTomcat> please. Because collection access code (I don't see any specific reason to use an array here)

**Comment by jstefl** [comment]
I added just one minor ask, as soon as pending comments are resolved I believe we can merge the code.

Thank you for working on this!

File: core/src/main/groovy/noe/tomcat/configure/SecureHttpConnectorTomcat.groovy (hunk 11 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/SecureHttpConnectorTomcat.groovy]: That's a good point.

**Comment by thanosploumis** [comment]
@jstefl Made the requested changes. My pleasure to work on some tool which all of us can use.

**Comment by thanosploumis** [comment]
Tested here : https://jws-jbcs-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/noe/job/noe-core/job/noe-core-integration-linux-clone-aploumis/49/

File: core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy [hunk 12]
```diff
@@ -63,6 +63,8 @@ class TomcatConfigurator {
    * have no unique id. *
    *
    * If there is not http (non-secure) connector then new connector is created.
+   * If there is already one, the connector attributes will be updated
```
[the remark this answers is not in the document]
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy:66]: We need to resolve this as well before merging.
I am still not sure about this one.

Imagine that you are new to `TomcatConfigurator` and you want to call code like
````
 ConnectorCertificateTomcat certificateObj = new ConnectorCertificateTomcat()
      .setCertificateFile(certFile)
      .setCertificateChainFile(chain)
      .setCertificateKeyAlias(keyAlias)
      .setCertificateKeyFile(keyFile)
      .setCertificateKeyPassword(keyPass)
      .setCertificateKeystoreFile(keystoreFile)
      .setCertificateKeystorePassword(keystorePass)
      .setCertificateKeystoreType(keystoreType)
      .setCertificateKeystoreProvider(keystoreProvider)
      .setCertificateType(certType)

    ConnectorSSLHostConfigTomcat sslHostConfObj = new ConnectorSSLHostConfigTomcat()
      .setCertificate(certificateObj)

    new TomcatConfigurator(tomcat)
      .httpsConnector(
        new SecureHttpConnectorTomcat()
          .setPort(testHttpsPort)
          .setSSLHostConfigs(sslHostConfObj)
      )
````

How you know, what configurations update the `server.xml` and what sets to specified values exactly? If you find out that some of the methods work differently than others (some sets, some updates), what would you think?

How do you suggest to handle this situation?

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy (hunk 5 above)
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy:121]: Please check the missing empty line in other submitted files as well.

**Comment by jstefl** [comment]
Thank you, I have added comments to pending comments, what needs to be resolved before merging.

File: core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy (hunk 12 above)
[the remark this answers is not in the document]
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy:66]: Hmmm I'm understanding what you mean
Maybe I should rephrase it to something like 

   `* If there is no secure HTTP connector then new secure HTTP connector is created.`
   `* If there is already one, its new attributes will be added along with the old ones`
   `* Inner elements have to be written to their exact format as a whole`
   
what do you think?

[the remark this answers is not in the document]
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy:66]: I am sure, that the text is good, as you suggesting. What I am not sure, that it will prevent surprises (not everybody reads the doc).
Any other idea?

[the remark this answers is not in the document]
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy:66]: I believe that someone not reading the docs, and not seeing the code-functionality will of course lead the user to surprises and it seems legit. From our side, I believe there is nothing more we can do :/ The naming is pretty good, java doc is present and the code is very clear. Nothing else comes to my mind right now

File: testsuite/src/test/groovy/noe/tomcat/BindingsTomcatConfiguratorIT.groovy [hunk 13]
```diff
@@ -385,6 +386,50 @@ abstract class BindingsTomcatConfiguratorIT extends TomcatTestAbstract {
...
+    String certPath = "path/to/cert"
+    String tType = "ttype"
+    String tFile = "tfile"
+    String tPass = "tpasswd"
+    String tProvider = "tprovider"
+    String protocols = "protocols"
+    String sslProtocol = "sslprotocol"
+    String ciphers = "several_ciphers"
+
+    Object sslHostConfObj = new ConnectorSSLHostConfigTomcat()
```
[the remark this answers is not in the document]
Comment (thanosploumis) [reply testsuite/src/test/groovy/noe/tomcat/BindingsTomcatConfiguratorIT.groovy]: Certificate test is included :)

File: core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy (hunk 5 above)
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/ConnectorSSLHostConfigTomcat.groovy]: Added the spaces.Seems i ve forgotten to push my commit

**Comment by thanosploumis** [comment]
I will add one more attribute inside SSLHostConfig which i find useful, named "hostname".Im changing the MR to WIP

**Comment by thanosploumis** [comment]
Tested lately here: https://jws-jbcs-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/noe/job/noe-core/job/noe-core-integration-linux-clone-aploumis/50/

File: core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy (hunk 12 above)
[the remark this answers is not in the document]
Comment (jstefl) [reply core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy:66]: Reading this again after a few weeks and to be doc more clear, it would imho help the following little change:

```
* If there is no secure HTTP connector, then a new secure HTTP connector is created.
* If there is already one, its new attributes will be added along with the old ones and old ones will be updated. Attributes are configured by methods with parameter type of java standard classes and primitive types.
* Inner elements have to be written to their exact form as a whole. Inner elements are configured by methods with custom classes.
```

And BTW, happy new year, let's hope that it will be better in every aspect!

[the remark this answers is not in the document]
Comment (thanosploumis) [reply core/src/main/groovy/noe/tomcat/configure/TomcatConfigurator.groovy]: Made the little change

Happy new year to you and your beloved ones Jan. Wishing the best :)

**Comment by thanosploumis** [comment]
@jstefl I was curious.Before making a new noe-core release.Is it possible to make one temporary-test version in order to run noe-tests on Jenkins with it?Because locally testing is kinda limited

**Comment by jstefl** [comment]
@thanosploumis yes it is. You can create a snapshot (so instead of command deploy you use command install) and install it into maven repository on NFS (so you need to be on machines what is connected on the correct NFS). Then in noe-test you will set to consume that snapshot instead of a regular version. IIRC it is possible to set it via parameters for noe-tests.

It is a very good idea to execute the noe-tests with a snapshot to see whether there are some issues with the new change. Especially, when MR introduces bigger changes. There are not many more painful things than broken tests during a test cycle. I highly recommended testing it also this way.

Please let me know how that went, and if all is as expected I will merge this MR.

**Comment by thanosploumis** [comment]
@jstefl Perfect. Let me get this straight.After the mvn install, i suppose inside target location will be created a snapshot.I will move it then to STATIC NFS. Then on noe-tests I will just edit the noe-core version to my snapshot and run it.
Im losing the step about moving the snapshot to static NFS. I suppose i will not connect via ssh and just moved it in its static NFS.It used be done from a central place which all machines are using.Have you any hint about this ?

**Comment by jstefl** [comment]
> Im losing the step about moving the snapshot to static NFS

You can ssh to any Jenkins machine (as hudson user). Checkout noe-core there and execute `mvn install`. It will deploy the snapshost into default maven repo (which is in the home directory of user hudson -> which is on NFS), just make sure, that you use the same repo when executing noe-tests (that the repo is not changed by some input arg). 

Alternatively, you can create a snapshot in your local machine and `scp` it to the maven target repo manually (just place it on the right place in the directory strucure).

**Comment by thanosploumis** [comment]
Jan sorry for continuing the questions, how can I ssh to jenkins machine if it is not online. ?In past, I made a test with sleep to make little changes during the machine was online but I know its not the right way and practice

**Comment by thanosploumis** [comment]
Sorry for asking again Jan. Is there a link with all nodes(IP's) available for connection? I remember the old days to using it, but now i cant find it

**Comment by thanosploumis** [comment]
Tested here: https://webservers-jenkins-csb-jwsqe.cloud.paas.psi.redhat.com/job/jws5-java-tests-RHEL-clone-aploumiss/28/

Hibernate-repo and embedded tomcat have minor change on repository and seems they did not find noe-core snapshot.However, i believe it is not need to rerun these profiles as we see all the rest profiles on all mediums/javas have their test passed except few expected failures.

**Comment by jstefl** [comment]
+1 for merge 

@AdamKrajcik @jan-onderka @PaulLodge @mmadzin do you have any additional comments, please?

**Comment by PaulLodge** [comment]
No comments from me +1 for the merge

**Comment by mmadzin** [comment]
Briefly went through it and I don't have any comments. 

+1 for the merge
